pennylane_lightning/core/bindings/Bindings_nb.hpp

```diff
@@ -0,0 +1,793 @@
+#include "Macros.hpp" // CPUArch
+#include "Memory.hpp" // alignedAlloc
+#include "Observables.hpp"
+#include "Util.hpp" // for_each_enum, PL_reinterpret_cast
+
+// Include backend-specific headers and define macros based on compile flags
+#ifdef _ENABLE_PLQUBIT
+#include "AdjointJacobianLQubit.hpp"
+#include "LQubitBindings_nb.hpp"
+#include "MeasurementsLQubit.hpp"
+#include "ObservablesLQubit.hpp"
+
+#define LIGHTNING_MODULE_NAME lightning_qubit_nb
+
+namespace {
+using namespace Pennylane::LightningQubit;
+using namespace Pennylane::LightningQubit::Algorithms;
+using namespace Pennylane::LightningQubit::Observables;
+using namespace Pennylane::LightningQubit::Measures;
+using namespace Pennylane::LightningQubit::NanoBindings;
+} // namespace
+
+#elif _ENABLE_PLKOKKOS == 1
+#include "AdjointJacobianKokkos.hpp"
+#include "LKokkosBindings_nb.hpp"
+#include "MeasurementsKokkos.hpp"
+#include "ObservablesKokkos.hpp"
+
+#define LIGHTNING_MODULE_NAME lightning_kokkos_nb
+
+namespace {
+using namespace Pennylane::LightningKokkos;
+using namespace Pennylane::LightningKokkos::Algorithms;
+using namespace Pennylane::LightningKokkos::Observables;
+using namespace Pennylane::LightningKokkos::Measures;
+using namespace Pennylane::LightningKokkos::NanoBindings;
+} // namespace
+
+#elif _ENABLE_PLGPU == 1
+#include "AdjointJacobianGPU.hpp"
+#include "BindingsCudaUtils.hpp"
+#include "LGPUBindings_nb.hpp"
+#include "MeasurementsGPU.hpp"
+#include "ObservablesGPU.hpp"
+
+#define LIGHTNING_MODULE_NAME lightning_gpu_nb
+
+namespace {
+using namespace Pennylane::LightningGPU;
+using namespace Pennylane::LightningGPU::Algorithms;
+using namespace Pennylane::LightningGPU::Observables;
+using namespace Pennylane::LightningGPU::Measures;
+using namespace Pennylane::LightningGPU::NanoBindings;
+} // namespace
+
+#elif _ENABLE_PLTENSOR == 1
+#include "AdjointJacobianTNCuda.hpp"
+#include "MeasurementsTNCuda.hpp"
+#include "ObservablesTNCuda.hpp"
+
+#define LIGHTNING_TENSOR_MODULE_NAME lightning_tensor_nb
+
+namespace {
+using namespace Pennylane::LightningTensor::TNCuda;
+using namespace Pennylane::LightningTensor::TNCuda::Observables;
+using namespace Pennylane::LightningTensor::TNCuda::Measures;
+using namespace Pennylane::LightningTensor::TNCuda::NanoBindings;
+} // namespace
+
+#else
+static_assert(false, "Backend not found.");
+#endif
+
+/// @cond DEV
+namespace {
+using Pennylane::Util::bestCPUMemoryModel;
+using Pennylane::Util::CPUMemoryModel;
+} // namespace
+/// @endcond
+
+namespace nb = nanobind;
+namespace Pennylane::NanoBindings {
+
+/**
+ * @brief Register applyMatrix.
+ */
+template <class StateVectorT>
+void registerMatrix(
+    StateVectorT &st,
+    const nb::ndarray<const std::complex<typename StateVectorT::PrecisionT>,
+                      nb::c_contig> &matrix,
+    const std::vector<std::size_t> &wires, bool inverse = false) {
+    using ComplexT = typename StateVectorT::ComplexT;
+    using Pennylane::Util::PL_reinterpret_cast;
+
```

Comment:
We should probably do a size check here, e.g. `assert(matrix.size() == 2^(wires.size()*2) )`